int32 current_state
int32 next_state
